[0016] docs: verification rfc (#80)
nervosnetwork/rfcs: rfcs/0016-verification/0016-verification.md

---
Number: "0016"
Category: Informational
Status: Draft
Author: Dingwei Zhang
Organization: Nervos Foundation
Created: 2019-03-08
---

# CKB-Verification

## Abstract

This RFC describes ckb block data verification rules.

Verification rules are split into three part in CKB implementation. `Header Verification`,  `Block Verification` and `Transaction Verification`.

## Specification

1. Header Verification

    1. proof-of-work verification

>         Block to be valid it must hash to a value less than the current target. current hash algorithm is Cuckoo Cycle.

Reviewer: ```suggestion
        The `seal` field in block header is miner's proof of work, it MUST satisfy the block's difficulty target. Cuckoo Cycle is the temporary proof-of-work function used in CKB testnet, a new hash function will be proposed and used in mainnet.
```

    2. number verification

>         CKB block header contain block number, it MUST equal parent’s block number incremented by one, the genesis block has a number of zero.

Reviewer: ```suggestion
        The `number` field in block header MUST equal parent’s block number incremented by one, the genesis block has a number of zero.
```

    3. block version verification

>         Block version MUST obey consensus rules.

Reviewer: This is too vague. This rfc is defining consensus rules.

You can say: Block version MUST be zero.

    4. timestamp verification

>         Timestamp (Unix’s time in milliseconds) of block timestamp MUST greater than the median timestamp of previous 11 blocks and less than the network-adjusted time + 15 seconds.

Reviewer 2: Please use a variable(`future block time`) to instead `15 second`, the value should be adjust after consensus paper released.

Reviewer: Use a variable, or a larger number before we have results from experiments.

    5. difficulty verification

>         The canonical difficulty of a block of header is defined as Diff == Max(Diff_last * o_last / o, Diff_last * 2) (o: orphan rate), It MUST greater or equal genesis' difficulty.

Reviewer: I think it's better to have a 'definition' section before specification, to define the exact meaning of concepts like 'orphan rate', 'block height' etc.

1. Block Verification

>    1. empty transactions verification

Reviewer: I think this rule is redundant with cellbase rule (block MUST include exactly one cellbase transaction).

        Block transactions MUST not be empty.

   2. duplicate

>         Block transactions MUST be unique.

Reviewer: This is vague because 'unique' is not defined. Try to express it in a more rigorous way.

   3. cellbase

      * position: cellbase MUST be the first transaction in block.

>       * input: cellbase MUST contain one null input, the input's script fill binary with current block number.

Reviewer: It's unnecessary to fill input's script with block number because we have it in block header.

Reviewer: any idea on above comment?

Author: https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki
It used to prevent intentional creation of duplicate cellbase transactions.

Reviewer 3: This ensures that no two transactions in the chain have the same content.

>       * output_capacity must less or equal consensus rules specify block reward.

Reviewer: ```suggestion
      * output_capacity must less or equal than block reward. Block reward includes primary issuance, secondary issuance and transaction fees, according to RFC#0015.
```

   4. merkle_root

>         Verify transaction [merkle-proof][1].

Reviewer: Too short. Please describe how is the merkle root formed briefly.

   5. transaction proposal rules

>         Commit transaction must be proposed within consensus rules specify proposal window.

Reviewer: Maybe use some placeholder like '<consensus rfc>' as reference here?

   6. uncles

      * uncles hash: The hash MUST match the [CFB][2] serialized uncles.

>       * number: there are at most 2 uncles included in one block.

Reviewer: Why 2 uncles at most? An inappropriate limit will affect the calculation of orphan rate.

Reviewer: any idea on above comment?

Author: uncle can be include in a window H(b) - 8 < H(u) < H(b). 
2 uncles at most for one block is not very tight.

Difficulty adjust has an upper limit( double currently ) for one epoch, so we collect uncles as much as we need, not as much as we can.
whether 2 is ok, we may need some test. 
Maybe use a variable for now?

      * depth: an uncle must be the k-th generation ancestor of Block, where 2 <= k <= 7.
      * epoch: uncle must be the same epoch with block.
      * uniqueness: an uncle must be different from all uncles included in previous blocks and all other uncles included in the same block (non-double-inclusion).

   7. block size limit

>         The maximum size in bytes that the consensus rules allow a block to be. The current block size limit is 10_000_000 bytes(10MB).

Reviewer: ```suggestion
        The block size MUST be less or equal than block size limit. The block size limit is 10_000_000 bytes (10MB).
```

   8. cycles limit

        The maximum cycles expenditure per block. The current cycles limit is 100_000_000.

2. Transaction Verification

>     Below transaction verification rules description exclude cellbase by default.

Reviewer: ```suggestion
    Cellbase transaction is excluded from following rules unless otherwise noted.
```

    1. transaction version verification

>         Transaction version MUST obey consensus rules.

Reviewer: This is vague.

    2. null verification

        All transaction inputs MUST not be null.

    3. empty inputs/outputs verification

        Transaction inputs or outputs MUST not empty.

    4. capacity verification

>        * Transaction capacity sum of inputs MUST greater or equal capacity sum of outputs.

Reviewer: ```suggestion
       * The total capacity referenced by transaction inputs MUST be greater or equal than the total capacity of outputs.
```

Need a definition for 'capacity' somewhere.

>        * All outputs' occupied_capacity MUST less than or equal to specify capacity.

Reviewer: ```suggestion
       * Each output's occupied_capacity MUST less than or equal than its capacity.
```

    5. input uniqueness verification

>         Input must be unique.

Reviewer: What does 'unique' mean here?

    6. input/dep status verification

>         All inputs and deps MUST be live cell.

Reviewer: What is a 'live cell'?

    7. script verification

>         Execution script in CKB-VM, perform user-define validation rules.

Reviewer: This is vague. A brief description and reference to CKB-VM is needed.

   4. transaction proposal rules

        Commit transaction must be proposed within consensus rules specify proposal window.

   5. uncles

      * uncles hash: The hash MUST match the [CFB][3] serialized uncles.
      * number: there are at most 2 uncles included in one block.
      * depth: we define uncle number H(u), block number H(b), uncle number MUST satisfy H(b) - m < H(u) < H(b). Currently, m = 8.
      * epoch: uncle MUST be the same epoch with block.
      * uniqueness: an uncle must be different from all uncles included in previous blocks and all other uncles included in the same block (non-double-inclusion).

   6. block size limit

        The block size MUST be less or equal than block size limit. The current block size limit is 10_000_000 bytes (10MB).

   7. cycles limit

        The maximum cycles expenditure per block. The current cycles limit is 100_000_000.

3. Transaction Verification

    Cellbase transaction is excluded from following rules unless otherwise noted.

    1. transaction version verification

        The version field is for doing parallel soft forking deployments. Currently, transaction version MUST be zero.

    2. null verification

>         All transaction inputs MUST not be null.

Reviewer 4: I'd suggest to emphasize the "not" part as well. 

Only make the "MUST" bold is really misleading on a first glance.

    3. empty inputs/outputs verification

>         Transaction inputs or outputs MUST not empty.

Reviewer 4: I'd suggest to emphasize the "not" part as well. 

Only make the "MUST" bold is really misleading on a first glance.
